### DISCUSSION TITLE
AAD authentication quirk

### DISCUSSION THREAD

**Discussion by RedSail2725** [datalust/seq-tickets#2011] [Help]
Hi folks,

I am setting up a Seq environment with Azure Active Directory authentication.  The documentation is straight forward enough, but when I activate it my login button only references localhost:

https://localhost.com:5341/oidc/challenge?toLocalUrl=%2Fsettings%2Fsystem%2Fauth

I have a fully qualified domain with associated certificate.  Local authentication against the full url works and my api.listenUris reflect the FQDN as well. (https://seq.example.com:5341)

Could someone point me to what I have missed?  Thanks!

**Comment by liammclennan** [comment]
Hi @RedSail2725 

Try setting `api.canonicalUri` to your Seq URI.

**Reply by RedSail2725** [reply]
Hi Liam,

Thanks for the quick reply.  That feels like the right answer, but I've got more odd behaviour in changing the setting.  I am using a wildcard certificate, which I think should be okay.

- When hitting https://seq.example.com:5341/ in my browser I get a login prompt if I run the following:
   seq config set -k api.cononicalUri -v https://localhost:5341/  
   
   So FQDN in the browser, localhost for the canonicalUri.
   
- If I set:  seq config set -k api.cononicalUri -v https://seq.example.com:5341/ and then visit the site in my browser I get **HTTP Error 503. The service is unavailable.**

Are wildcards supported?

**Reply by nblumhardt** [reply]
Hi! It sounds like there's a mismatch going on somewhere, here's what I think you will need to do:

```
seq config set -k api.listenUris -v https://seq.example.com:5341
seq config set -k api.canonicalUri -v https://seq.example.com:5341
seq bind-ssl -p 5341 --thumbprint="THUMBPRINT"
seq service restart
```

(Where `THUMBPRINT` is your certificate thumbprint.)

The `bind-ssl` command is fallible, it's best to watch its output carefully.

If the above combination doesn't get things working, checking `C:\ProgramData\Seq\Logs` might turn up some error messages.

Looking forward to hearing how you go,

Best regards,
Nick

**Reply by RedSail2725** [reply]
It's interesting, the only way I can get the page to load with the correct link is: 

**seq config set -k api.listenUris -v https://seq.example.com:5341,https://localhost:5341** (so two Uris)

I don't have any other references that I can see to localhost, I'm not referencing it in my browser session and it doesn't show in the json or seq config output from the command line.

DNS resolves cleanly and I have even tested another certificate.  There is no proxy.

Without it I get a 503, with it I can authenticate and get redirected (I have the successful Azure logs), but I see this in the browser:

OpenID Connect authentication failed
No extended description available.

Server logs show JWT validation errors, which makes sense if it's referencing localhost in the token.

I agree that this should work:

seq config set -k api.listenUris -v https://seq.example.com:5341
seq config set -k api.canonicalUri -v https://seq.example.com:5341
seq bind-ssl -p 5341 --thumbprint="THUMBPRINT"
seq service restart

It just doesn't.  I even get the same behaviour from a browser session directly from the desktop of the server itself.

I must be missing something very basic that requires the localhost entry, any chance you have output from a valid seq config that I can use to compare?

**Reply by nblumhardt** [reply]
Thanks for your reply. If the above didn't work, there are a few more possibilities to check out.

First, in the app registration for Seq in AAD, do you have only one "redirect uri", and is it `https://seq.example.com:5341/aad`?

Next, in the "working" configuration that fails on redirect, what is the scheme, hostname, and port you see when you're redirected back to Seq (where the error message is displayed?).

Finally, the issue with the "correct but not working" configuration might be permissions-related; to check and fix that, run Start > Seq Service Administration, and then click through supplying only the exact values used in the configuration sample I gave above, that is, https://seq.example.com:5341 as the listen URI, and the thumbprint of your certificate (no localhost entries, etc.).

Hopefully one of these will show up the issue. Fingers crossed 🤞

**Reply by RedSail2725** [reply]
Hi Nick,

Thanks for the reply. I can get authenticated and redirected back to the application but I'm still seeing the same "no extended description available".  If I look in AAD I see the successful session. This is from the logs (bottom).

I pulled token IDs out of the output since this is a public forum and I'd rather not share the URLs for the same reason.  We are inside of a support agreement, though, is it possible to set up a call?  I feel as if we're close and can probably resolve the issue in a couple of minutes.

'The number of keys' warning seems pretty specific, but I wasn't able to find anything in the forum.

Thanks for the help!

---------------------------- 

{"@t":"2023-11-06T18:38:14.3191442Z","@mt":"OpenID Connect authentication failed with {Error} / {ErrorDescription}","@l":"Error","@x":"Microsoft.IdentityModel.Tokens.SecurityTokenSignatureKeyNotFoundException: IDX10501: Signature validation failed. Unable to match key: \nkid: '[redacted, didn't want to post the key publicly]'. \nNumber of keys in TokenValidationParameters: '4'. 

**Number of keys in Configuration: '0'.** 

Exceptions caught:
 '[PII of type 'System.Text.StringBuilder' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'. \
 
 token: '[PII of type 'System.IdentityModel.Tokens.Jwt.JwtSecurityToken' is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'.
 
 at Microsoft.IdentityModel.Tokens.InternalValidators.ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(SecurityToken securityToken, Nullable`1 notBefore, Nullable`1 expires, String kid, TokenValidationParameters validationParameters, BaseConfiguration configuration, StringBuilder exceptionStrings, Int32 numKeysInConfiguration, Int32 numKeysInTokenValidationParameters)
 
 at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateSignature(String token, JwtSecurityToken jwtToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
 
 at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateSignatureAndIssuerSecurityKey(String token, JwtSecurityToken jwtToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)

 at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateJWS(String token, TokenValidationParameters validationParameters, BaseConfiguration currentConfiguration, SecurityToken& signatureValidatedToken, ExceptionDispatchInfo& exceptionThrown)
 
 --- End of stack trace from previous location
 
 at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateToken(String token, JwtSecurityToken outerToken, TokenValidationParameters validationParameters, SecurityToken& signatureValidatedToken)
 
 at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateToken(String token, TokenValidationParameters validationParameters, SecurityToken& validatedToken)
 
 at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.ValidateToken(String idToken, AuthenticationProperties properties, TokenValidationParameters validationParameters, JwtSecurityToken& jwt)

 at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleRemoteAuthenticateAsync()","Error":"No error provided","ErrorDescription":"No extended description available","RequestId":"[redacted, didn't want to paste in a public forum]","RequestPath":"/aad"}

**Reply by nblumhardt** [reply]
Thanks for the follow-up @RedSail2725. I'm not sure why it would impact you and not others (perhaps a more sophisticated AAD configuration/non-default Azure setup?) but it's looking like the problem described in this post:

https://stackoverflow.com/questions/58856735/idx10501-signature-validation-failed-unable-to-match-keys

We'll investigate and let you know what we can come up with 👍

**Reply by nblumhardt** [reply]
Sorry about the hold-up on this, I'll loop back shortly with an update.

**Comment by RedSail2725** [comment]
Thanks :-)  

As a stop gap, is there any documentation available for separating the ingestion from the admin ports?

**Reply by nblumhardt** [reply]
```
seq config -k api.ingestionPorts -v 12345
seq service restart
```

sets a restriction so that port 12345 can receive ingestion only; the port still needs to appear in an `api.listenUris` entry in order for Seq to serve it. You should check by hand that the restriction has been successful before exposing the port on the basis of this setting, though.

I'm assuming that you mean you'll also employ username/password authentication in the meantime (as part of your stopgap solution), please let me know if I'm misunderstanding.

Hope this helps,
Nick
